lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -832,6 +1012,50 @@ std::string SPIRVToOCLBase::getGroupBuiltinPrefix(CallInst *CI) {
   return Prefix;
 }
 
+std::string
+SPIRVToOCLBase::getOCLImageOpaqueType(SmallVector<std::string, 8> &Postfixes) {
+  SmallVector<int, 7> Ops;
+  for (unsigned I = 1; I < 8; ++I)
+    Ops.push_back(atoi(Postfixes[I].c_str()));
+  SPIRVTypeImageDescriptor Desc(static_cast<SPIRVImageDimKind>(Ops[0]), Ops[1],
+                                Ops[2], Ops[3], Ops[4], Ops[5]);
+
+  std::string OCLStructName =
+      std::string(kSPR2TypeName::OCLPrefix) + rmap<std::string>(Desc);
+
+  SPIRVAccessQualifierKind Acc = static_cast<SPIRVAccessQualifierKind>(Ops[6]);
+  insertImageNameAccessQualifier(Acc, OCLStructName);
+  return OCLStructName;
+}
+
+void SPIRVToOCLBase::translateOpaqueTypes() {
+  for (auto *S : M->getIdentifiedStructTypes()) {
+    StringRef STName = S->getStructName();
+    bool IsSPIRVOpaque =
+        S->isOpaque() && STName.startswith(kSPIRVTypeName::PrefixAndDelim);
+
+    if (!IsSPIRVOpaque)
+      continue;
+
+    SmallVector<std::string, 8> Postfixes;
+    std::string DecodedST = decodeSPIRVTypeName(STName, Postfixes);
+
+    if (!SPIRVOpaqueTypeOpCodeMap::find(DecodedST))
+      continue;
+
+    Op OP = SPIRVOpaqueTypeOpCodeMap::map(DecodedST);
+    std::string OCLOpaqueName;
+    if (OP == OpTypeImage)
+      OCLOpaqueName = getOCLImageOpaqueType(Postfixes);
+    else if (isSubgroupAvcINTELTypeOpCode(OP))
+      OCLOpaqueName = OCLSubgroupINTELTypeOpCodeMap::rmap(OP);
+    else
+      continue;
```

Comment:
Yeah, we will need this transformation for other opaque types as well, for example queue or event.
I'm going to base my patch related to it on these changes.
@AlexeySotkin , could you please review this patch so it will be more convenient?